st2common/st2common/services/workflows.py

```diff
@@ -1129,3 +1129,57 @@ def update_execution_records(wf_ex_db, conductor, update_lv_ac_on_statuses=None,
     if status_changed and wf_lv_ac_db.status in ac_const.LIVEACTION_COMPLETED_STATES:
         LOG.info('[%s] Workflow action execution is completed and invoking post run.', wf_ac_ex_id)
         runners_utils.invoke_post_run(wf_lv_ac_db)
+
+
+@retrying.retry(retry_on_exception=wf_exc.retry_on_exceptions)
+def request_rerun(ac_ex_db, st2_ctx, options):
+    ac_ex_id = str(ac_ex_db.id)
+    LOG.info('[%s] Processing rerun request for workflow.', ac_ex_id)
+
+    wf_ex_id = st2_ctx['workflow_execution_id']
+    wf_ex_db = wf_db_access.WorkflowExecution.get_by_id(wf_ex_id)
+
+    if not wf_ex_db:
+        raise wf_exc.WorkflowExecutionNotFoundException(ac_ex_id)
+
+    if wf_ex_db.status != statuses.FAILED:
+        msg = '[%s] Workflow execution "%s" is not rerunable because its status is not "%s".'
+        LOG.info(msg, ac_ex_id, wf_ex_id, statuses.FAILED)
+        return
+
+    if wf_ex_db.status in statuses.RUNNING_STATUSES:
+        msg = '[%s] Workflow execution "%s" is not rerunable because it is already active.'
+        LOG.info(msg, ac_ex_id, ac_ex_id)
+        return
+
+    wf_ex_db.result = {}
+    wf_ex_db.error = {}
+    wf_ex_db.action_execution = ac_ex_id
+    wf_ex_db.context['st2'] = st2_ctx['st2']
+    wf_ex_db.context['parent'] = st2_ctx['parent']
+
+    conductor = deserialize_conductor(wf_ex_db)
+    conductor.request_workflow_rerun(options)
+    data = conductor.serialize()
+    status = data['state']['status']
+
+    if status in statuses.COMPLETED_STATUSES:
+        raise wf_exc.WorkflowExecutionIsCompletedException(str(wf_ex_db.id))
+
+    if status != statuses.RESUMING:
+        msg = '[%s] Workflow execution "%s" is not reran because it is not running status.'
+        LOG.info(msg, ac_ex_id, str(wf_ex_db.id))
+        return
```

Comment:
Instead of LOG.info and returning here, can we throw an exception?  Create a new workflow exception named `WorkflowExecutionRerunException` that output the error message `[%s] Unable to rerun workflow execution. The workflow execution is in "%s" status.`.